### DISCUSSION TITLE
VSCode Debug Configuration in Monorepos

### DISCUSSION THREAD

**Discussion by philipj93** [vercel/turborepo#2639] [Help]
It can be super helpful to have the debugger configured in VSCode for debugging Next.js applications. The Next.js docs [recommend a configuration](https://nextjs.org/docs/advanced-features/debugging#debugging-with-vs-code) that can be added to the `.vscode/launch.json` file to make it easy to debug projects with breakpoints. 

It would be great to be able to debug multiple applications running at the same time (e.g. a docs site and a dashboard using UI components). This would probably be a useful addition to the Turborepo docs, since most users are likely using VSCode. However, it's unclear how/if this works with Turborepo. 

I've just started to explore the options here, but haven't really got anywhere. Here's a couple of the questions that have come up so far:

- Does each application in a monorepo need its own `.vscode/launch.json` file configured, or would a single debug configuration at the project root suffice?
- Does one need to add a specific task to the `turbo.json` file for debugging?

Has anyone been able to set up a working debugger configuration in their monorepo? I'd love to see how I could get this set up for my own projects and possibly contribute to the docs to help others get started as well.

**Comment by AndyOooh** [comment]
I'm currently trying to find out about this as well.

Posted a new question before finding this:
https://github.com/vercel/turbo/discussions/4860

Did you ever find a solution?

**Comment by ethanpowersMOD** [comment]
I'm looking for this as well.

**Comment by artem-alek** [comment]
I am still looking too...

**Comment by aaronhawkey** [comment]
I am using the recommended NextJS configuration for the vscode debugger and I am having issues with breakpoints working within my packages. Is this related?

**Comment by kimzeevaarders** [comment]
Same problem here as well, since migration to turborepo debugging not working anymore...

**Comment by markomitranic** [comment]
This is a pretty common thing to do in other ecosystems and on larger projects. Sometimes its a monorepo, other times 80 microservice repos on a minikube - the principle is the same - only run what you must in `debug` mode, run everything else as you normally would.

- Start the `turbo dev` or `turbo start` tasks as you normally would.
- Don't start them all, filter out the services you wish to debug `turbo dev -F '!@acme/web'`. *Note: the `'` are important to prevent Bash from interpreting the ! character as a history expansion event and allow Turbo to correctly process the negative filter.*
- Now, run the VSCode debug configuration only for the services you actually wish to debug:
```json
// .vscode/launch.json
    {
      "name": "Next.js: debug full stack",
      "type": "node-terminal",
      "request": "launch",
      "command": "npm run dev",
      "cwd": "${workspaceFolder}/apps/web",
      "serverReadyAction": {
        "pattern": "started server on .+, url: (https?://.+)",
        "uriFormat": "%s",
        "action": "debugWithChrome"
      }
    }
```

If you need to run more than 1 debugger, VSCode is capable to run tasks in parallel, although it will get confusing soon. The important part of the launch config above is `"cwd": "${workspaceFolder}/apps/web"` which sets the work directory. You might think that running `npm run -w apps/web dev` should technically work similarly to changing the `cwd`, but in practice, setting the `cwd` explicitly ensures that all path resolutions and VSCode's environment variables when attaching to the debug session are correctly aligned with the workspace directory. 🤷‍♂️ 

**Bonus points:**

Since Next.JS is a true full-stack framework, we shouldn't forget CLI script debugging. Turborepo excells at having us decouple administrative scripts from our applications, which we usually place either in `/packages/scripts/` or into `/tooling/scripts/`. The principle is the same, and you don't need turbo for it.

- You run whatever supporting services you need to run via Turbo, as you normally would.
- This time no need to filter out anything, as CLI scripts don't have a `dev` and wouldn't run anyway.
- Run the following VSCode launch config:

```json
// .vscode/launch.json
    {
      "type": "node-terminal",
      "request": "launch",
      "name": "Debug CLI Script",
      "command": "npm run cli src/example.ts",
      "cwd": "${workspaceFolder}/tooling/scripts",
      "skipFiles": ["<node_internals>/**"]
    }
```

**Comment by goshantmeher** [comment]
This is my working launch.json file with NestJs ,NextJs setup
file: **.vscode/launch.json**
```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Debug API",
      "type": "node",
      "request": "launch",
      "runtimeExecutable": "pnpm",
      "runtimeArgs": ["run", "start:debug"],
      "cwd": "${workspaceFolder}/apps/api"
    },
    {
      "name": "Debug Web",
      "type": "node-terminal",
      "request": "launch",
      "command": "pnpm run dev",
      "cwd": "${workspaceFolder}/apps/web"
    }
  ],
  "compounds": [
    {
      "name": "Debug All",
      "configurations": ["Debug API", "Debug Web"]
    }
  ]
}
```

**Comment by gethari** [comment]
## Is there a working solution as of 2025 ?

**Comment by Ethan-telenor** [comment]
I think I made it work - try this and let me know if this doesn't work

### important points
- `type: node` instead of `node-terminal`
  - avoids VS Code injecting its bootloader path, so the Node process starts cleanly and the debugger can attach reliably.
- `runtimeArgs: ["--inspect=PORT"]`
  - this starts the Node inspector on that port; VS Code’s Node debug adapter knows to attach to that port automatically, just like any other Node process launched with `--inspect`.
  - **Each Next dev server needs its own --inspect port, and VS Code needs a separate debug config per target to attach.**
- `program` pointing to each app’s local Next binary (e.g., `${workspaceFolder}/apps/<app>/node_modules/next/dist/bin/next`)
  - with pnpm, Next lives in the app’s own `node_modules`, not hoisted to the root; using the app-local binary prevents “module not found” and ensures the right Next version/paths.
- `cwd` set to the app folder: aligns module resolution, env, and path expectations with the app context; avoids mismatches that happen when running from the repo root.

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Dev: fe1 (Next + Turbopack)",
      "runtimeExecutable": "node",
      "runtimeArgs": ["--inspect=9201"],
      "program": "${workspaceFolder}/apps/fe1/node_modules/next/dist/bin/next",
      "args": ["dev", "--turbopack", "-p", "3001"],
      "cwd": "${workspaceFolder}/apps/fe1",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen",
      "presentation": { "hidden": true },
      "skipFiles": ["<node_internals>/**"]
    },
    {
      "type": "node",
      "request": "launch",
      "name": "Dev: fe2 (Next + Turbopack)",
      "runtimeExecutable": "node",
      "runtimeArgs": ["--inspect=9202"],
      "program": "${workspaceFolder}/apps/fe2/node_modules/next/dist/bin/next",
      "args": ["dev", "--turbopack", "-p", "3002"],
      "cwd": "${workspaceFolder}/apps/fe2",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen",
      "presentation": { "hidden": true },
      "skipFiles": ["<node_internals>/**"]
    },
    {
      "type": "node",
      "request": "launch",
      "name": "Dev: be1 (Next + Turbopack)",
      "runtimeExecutable": "node",
      "runtimeArgs": ["--inspect=9210"],
      "program": "${workspaceFolder}/apps/be1/node_modules/next/dist/bin/next",
      "args": ["dev", "--turbopack", "-p", "4000"],
      "cwd": "${workspaceFolder}/apps/be1",
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen",
      "presentation": { "hidden": true },
      "skipFiles": ["<node_internals>/**"]
    }
  ],
  "compounds": [
    {
      "name": "Dev: all apps (fe1 + fe2 + be1)",
      "configurations": [
        "Dev: fe1 (Next + Turbopack)",
        "Dev: fe2 (Next + Turbopack)",
        "Dev: be1 (Next + Turbopack)"
      ],
      "stopAll": true
    }
  ]
}
```
